swc/private/swc.bzl

```diff
@@ -0,0 +1,88 @@
+"Internal implementation details"
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+_attrs = {
+    "srcs": attr.label_list(allow_files = True, mandatory = True),
+    "args": attr.string_list(),
+    "source_maps": attr.string(),
+    "data": attr.label_list(default = [], allow_files = True),
+    "swcrc": attr.label(allow_single_file = True),
+    "swc_cli": attr.label(
+        default = "@aspect_rules_swc//swc:cli",
+        executable = True,
+        cfg = "exec",
+    ),
+}
+
+_outputs = {
+    "js_outs": attr.output_list(),
+    "map_outs": attr.output_list(),
+}
+
+def _impl(ctx):
+    outputs = ctx.outputs.js_outs + ctx.outputs.map_outs
+    outputs_depset = depset(outputs)
+    source_maps = len(ctx.outputs.map_outs) > 0
+
+    for src in ctx.files.srcs:
+        js_out = ctx.actions.declare_file(paths.replace_extension(src.basename, ".js"), sibling = src)
+        outs = [js_out]
+        if source_maps:
+            outs.append(ctx.actions.declare_file(paths.replace_extension(src.basename, ".js.map"), sibling = src))
+        args = ctx.actions.args()
+
+        # Add user specified arguments *before* rule supplied arguments
+        args.add_all(ctx.attr.args)
+
+        # Pass in the swcrc config if it is set; if it is
+        # then source paths are expected to be relative to the swcrc directory
+        src_path = src.path
+        if ctx.file.swcrc:
+            swcrc_path = ctx.file.swcrc.path
+            swcrc_directory = paths.dirname(swcrc_path)
+            args.add_all([
+                "--config",
+                swcrc_path,
+            ])
+            if not src_path.startswith(swcrc_directory):
+                fail("sources must be in swcrc directory or subdirector if swcrc is specified")
```

Comment:
this `fail` may not be needed anymore since `paths.relativize` below should provide the correct path; need to test

a "TODO: check if this is true" at least so we don't forget?

Reply:
commented the condition out for now with comment, easier to detect that it's missing than that it's unneeded

Reply:
👍